File: src/test/java/it/unibo/crabinv/Model/wave/WaveTest.java
package it.unibo.crabinv.Model.wave;

import it.unibo.crabinv.Model.entities.enemies.Enemy;
import it.unibo.crabinv.Model.entities.enemies.EnemyFactory;
import it.unibo.crabinv.Model.entities.enemies.EnemyType;
import it.unibo.crabinv.Model.entities.enemies.rewardService.RewardsService;
import it.unibo.crabinv.Model.entities.enemies.wave.Wave;
import it.unibo.crabinv.Model.entities.enemies.wave.WaveImpl;
import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.concurrent.atomic.AtomicBoolean;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.Mockito.*;

class WaveTest {

    @Test
    void waveShouldFinishAfterAllSpawnedEnemiesDie() {
        final EnemyFactory enemyFactory = mock(EnemyFactory.class);
        final RewardsService rewardsService = mock(RewardsService.class);

        final AtomicBoolean alive = new AtomicBoolean(true);
        final Enemy enemy = mock(Enemy.class);
        when(enemy.isAlive()).thenAnswer(inv -> alive.get());

        when(enemyFactory.createEnemy(
                any(EnemyType.class),
                anyDouble(),
                anyDouble(),
                anyDouble(),
                anyDouble()
        )).thenReturn(enemy);

        final EnemyType type = EnemyType.values()[0];

        double spawnYNorm   = 0.2;
        double bottomYNorm  = 0.8;

        final Wave wave = new WaveImpl(
                List.of(type),
                List.of(0),
                enemyFactory,
                rewardsService,
                5,
                spawnYNorm,
                bottomYNorm
        );

        // 1) primo tick: spawna 1 nemico, quindi non può essere finita
        wave.tickUpdate();
        assertFalse(wave.isWaveFinished());
        assertEquals(1, wave.getAliveEnemies().size());

        // 2) il nemico muore "nel gioco"
        alive.set(false);

        // 3) tick successivo: cleanup lo rimuove e la wave finisce
        wave.tickUpdate();
        assertTrue(wave.isWaveFinished());
        assertTrue(wave.getAliveEnemies().isEmpty());

        verify(rewardsService, times(1)).rewardEnemyDeath(enemy);
    }
}

File: src/test/java/it/unibo/crabinv/model/wave/WaveTest.kt
package it.unibo.crabinv.model.wave

import it.unibo.crabinv.model.entities.enemies.Enemy
import it.unibo.crabinv.model.entities.enemies.EnemyFactory
import it.unibo.crabinv.model.entities.enemies.EnemyType
import it.unibo.crabinv.model.entities.enemies.rewardservice.RewardsService
import it.unibo.crabinv.model.entities.enemies.wave.Wave
import it.unibo.crabinv.model.entities.enemies.wave.WaveImpl
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import org.mockito.ArgumentMatchers
import org.mockito.Mockito
import org.mockito.invocation.InvocationOnMock
import org.mockito.stubbing.Answer
import java.util.concurrent.atomic.AtomicBoolean

internal class WaveTest {
    @Test
    fun waveShouldFinishAfterAllSpawnedEnemiesDie() {
        val enemyFactory = Mockito.mock(EnemyFactory::class.java)
        val rewardsService = Mockito.mock(RewardsService::class.java)

        val alive = AtomicBoolean(true)
        val enemy = Mockito.mock(Enemy::class.java)
        Mockito.`when`(enemy.isAlive()).thenAnswer(Answer { _: InvocationOnMock? -> alive.get() })

        Mockito
            .`when`(
                enemyFactory.createEnemy(
                    ArgumentMatchers.any(EnemyType::class.java),
                    ArgumentMatchers.anyDouble(),
                    ArgumentMatchers.anyDouble(),
                    ArgumentMatchers.anyDouble(),
                    ArgumentMatchers.anyDouble(),
                ),
            ).thenReturn(enemy)

        val type = EnemyType.entries[0]

        val spawnYNorm = 0.2
        val bottomYNorm = 0.8

        val wave: Wave =
            WaveImpl(
                listOf<EnemyType?>(type),
                mutableListOf<Int?>(0),
                enemyFactory,
                rewardsService,
                5,
                spawnYNorm,
                bottomYNorm,
            )

        wave.tickUpdate()
        Assertions.assertFalse(wave.isWaveFinished())
        Assertions.assertEquals(1, wave.getAliveEnemies().size)

        alive.set(false)

        wave.tickUpdate()
        Assertions.assertTrue(wave.isWaveFinished())
        Assertions.assertTrue(wave.getAliveEnemies().isEmpty())

        Mockito.verify(rewardsService, Mockito.times(1)).rewardEnemyDeath(enemy)
    }
}
